sphinx_immaterial/apidoc_formatting.py

```diff
@@ -50,33 +52,29 @@ def visit_desc(self, node: sphinx.addnodes.desc) -> None:
         node["classes"].append("objdesc")
         super().visit_desc(node)
 
-    def visit_desc_type(self, node: sphinx.addnodes.desc_type) -> None:
+    def visit_desc_type(self, node: docutils.nodes.Element) -> None:
         self.body.append(
             self.starttag(node, tagname="span", suffix="", CLASS="desctype")
         )
 
-    def depart_desc_type(self, node: sphinx.addnodes.desc_type) -> None:
+    def depart_desc_type(self, node: docutils.nodes.Element) -> None:
```

Comment:
Its a shame we can't use the more specific type for specialized functions like these.

Reply:
Yeah.  I guess # type: ignore is the alternative but I'm not sure that is better

Reply:
Could we do a `Union[docutils.nodes.Element, <specific node type>]` instead of `type: ignore`?

Reply:
I suppose, but technically that union type should be equivalent to just `docutils.nodes.Element` so it doesn't add anything.

Reply:
I think I understand the need for this change. running `mypy --show-error-codes` on main branch yielded
```
sphinx_immaterial\apidoc_formatting.py:58: error: Argument 1 of "depart_desc_type" is incompatible with supertype "HTML5Translator"; supertype defines the argument type as "Element"  [override]
sphinx_immaterial\apidoc_formatting.py:58: note: This violates the Liskov substitution principle
sphinx_immaterial\apidoc_formatting.py:58: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```
The error is misleading because it says argument 1 (self) should be what the super class' argument 2 (node) type is. Errors like these are why I tend to avoid using mypy. Remember that mypy is still in "beta" (v0.950 is latest), so we'll have to take the advice from mypy output with a inquisitive mindset.

We could start a mypy config in pyproject.toml that ignores certain errors for certain files.
```
[mypy]

[mypy-sphinx_immaterial.apidoc_formatting]
disable-error-code = "override"
```

Reply:
It seems that mypy doesn't support ignoring certain errors per module. It can only be done globally, and I think it can only be done to 1 error code in the config file. It is starting to look like `type: ignore` is the only mechanism we can use to ignore an error, and I was hoping to avoid having that comment littered all over the src code.

The problem with `type: ignore` is that it is distracting and can cause violation of line length restrictions.

I still prefer to have the specific type in these specialized functions because it better describes some of the attributes implemented by a derivative of `docutils.nodes.Element`.

Reply:
As far as the argument numbers, I think it is just excluding `self`.

I agree that mypy isn't great (the typescript type system is much, much better).  In putting together this PR I discovered a few incorrect type annotations but no actual bugs.  Nonetheless the type annotations themselves certainly are useful for documentation, and mypy is needed to ensure they bear some resemblance to reality.

I'm reluctant to disable error codes file-wide, but I'll go ahead and change the `# type: ignore` to `#type: ignore[code]` so that it is more selective and documents what is being ignored.

Reply:
I'm looking into alternatives to mypy. I don't want to be tied down to a beta tool while this repo is still also in beta (too many bleeding edges).

I see that [microsoft has a tool called pyright](https://github.com/microsoft/pyright) that is available from npm. Since the type checking I've seen in VS Code (which uses pyright) is much more tolerant than mypy, I'm seriously considering the switch.

Reply:
I'm sorry you did all this work already. I don't like shooting down PRs, but this is something I would've liked to discuss in an issue first.

Reply:
There is `pytype`, which we use internally at Google.  However, it isn't really well supported for external use.  From what I've seen, mypy is pretty universal, and while it may be in "beta" it has been around for many years at this point.  However I haven't looked at `pywright` -- maybe it is a better alternative.

Reply:
I'm happy to put this on hold.

Reply:
I really like pyright in VS Code. There's few instances where I have to explicitly declare an object's type. I'm curious what IDE you use.

My main concern is we'd be losing specificity (in this instance) just to please the mypy tool. other errors like dynamic attribute definitions shouldn't require a `type: ignore` comment since it basically ignores an advantage to using python in general.

Reply:
I use Emacs --- it supports the language server protocol, though for Python I don't use one.

Have you tried running pyright on this package?

Dynamic attributes are tricky, because if you just misspell an attribute you would want to receive an error.

The issue with the `visit_xxx` members is that technically the error is correct, it is just that the type annotations in Sphinx aren't as precise as we would like so we are trying to override them.  Still we could use a `# type: ignore` there rather than change the types.

Reply:
> it is just that the type annotations in Sphinx aren't as precise as we would like

Agreed. I still don't want to extend that seemingly bad practice though (I'm sure sphinx devs have their reasons).

> Have you tried running pyright on this package?

I just ran it using `npx pyright .\sphinx_immaterial\*.py` and it found 25 source files for which it summarized:
```
Found 25 source files
... much better errors/warnings compared to mypy
242 errors, 1 warning, 0 informations
Completed in 3.353sec
```
It also made notice of the lack of configuration (in pyproject.toml) before file discovery.

Reply:
>> Could we do a `Union[docutils.nodes.Element, <specific node type>]` instead of type: ignore?
>
> I suppose, but technically that union type should be equivalent to just docutils.nodes.Element so it doesn't add anything.

I went forward with this idea because type **hinting** isn't just programmatic when reviewing (since the functions names are subject to authors' thought process).

resolving this long thread.